src/Command/CreateArticleResult.php

```diff
@@ -0,0 +1,23 @@
+<?php
+
+namespace App\Command;
+
+use App\Entity\Article;
+
+class CreateArticleResult implements CommandResult
```

Comment:
I guess that this command is supposed to represent what is returned by the command handler? If that’s the case, I’d say that it would be semantically clearer to choose another term than «command» for this kind of things, as a command normally induces a mutation.

Also, personally I’d keep things simple and not overly abstract here, and I would not implement any interface for this output DTO either. I don’t see how you could make such a contract flexible enough anyway: the `getResult()` method would have to have a `mixed` return type, which would make the return type actually pointless…

Reply:
I get your point, but in that case, maybe we can just return a domain object or iterable (for collections results) directly from the command handler, rather than to encapsulate it in a command result ?

Reply:
That’s the way I would personally go indeed, but some people would tell you that you should rather build a specific output DTO from the entity, and return that DTO instead. That would mean than only the domain layer is allowed to manipulate entities (and thus, repositories). That’s called the View Model pattern. In my opinion, it over-complicates things for not much of a benefit, so I generally don’t use it. This pattern is supposed to address the following problem: an entity generally exposes read-write capabilities, which is good in the context of the domain layer (command handlers in that case), but not so good when you want to present your entity in your view layer, because you don’t want your view layer to be able to mutate your entities. That’s where the view model comes into play. Personally, I’m fine with using entities in the view layer, but it generally requires 2 things:

* Keep the entity classes (hence, the model) simple, and do not use overly-complicated data structures (deep-nesting relationships, stuff like that), otherwise you’ll likely encounter problems with (de-)serialization and stuff like that
* Build your layers so that it’s impossible that any changes potentially applied to an entity in the view layer (which should be implicitly forbidden BTW) will never be persisted; in practice, it’s challenging to enforce this constraint at the code level, even though theoretically possible, but we often rely on the fact that the team will behave responsibly and apply good practices

Reply:
Thanks for that very complete answer sir :tophat: